File: 1/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define N 1000

void solve_file(char* filename) {
  int n_lines = 0;
  char line[N] = {0};
  int numbers[N] = {0};

  printf("Solving for file: %s\n", filename);
  FILE* infile = fopen(filename, "r");
  while(!feof(infile)) {
    fgets(line, N, infile);
    line[N] = '\0';
    numbers[n_lines] = atoi(line);
    n_lines++;
  }
  fclose(infile);

  for (int i = 0; i < n_lines; i++)
    for (int j = 0; j < n_lines; j++)
      if (numbers[i] + numbers[j] == 2020)
      {
        printf("%d\n", numbers[i] * numbers[j]);
        return;
      }

}

int main(int argc, char* argv[]) {
  for (int i = 1; i < argc; i++) {
    solve_file(argv[i]);
  }
  return 0;
}

File: 1/main.nim
const
  N* = 1000

proc solve_file*(filename: cstring) =
  var n_lines: cint = 0
  var line: array[N, char] = [0]
  var numbers: array[N, cint] = [0]
  printf("Solving for file: %s\n", filename)
  var infile: ptr FILE = fopen(filename, "r")
  while not feof(infile):
    fgets(line, N, infile)
    line[N] = '\x00'
    numbers[n_lines] = atoi(line)
    inc(n_lines)
  fclose(infile)
  var i: cint = 0
  while i < n_lines:
    var j: cint = 0
    while j < n_lines:
      if numbers[i] + numbers[j] == 2020:
        printf("%d\n", numbers[i] * numbers[j])
        return
      inc(j)
    inc(i)

proc main*(argc: cint; argv: ptr cstring): cint =
  var i: cint = 1
  while i < argc:
    solve_file(argv[i])
    inc(i)
  return 0
